File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/ui/AdaptersCallScreens.kt
package com.perrystreet.woof.konsist.ui

import com.lemonappdev.konsist.api.ext.list.withNameEndingWith
import com.lemonappdev.konsist.api.ext.list.withoutName
import com.perrystreet.woof.konsist.Assertions.assertTrue
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.KonsistUtils.body
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class AdaptersCallScreens : BehaviorSpec() {
    init {
        Given("A feature Adapter composable") {
            val adapters = KonsistUtils.featureModules.functions().withNameEndingWith("Adapter").withoutName("ErrorAdapter")

            Then("It renders a Screen composable") {
                adapters.assertTrue(message = Message) { adapter -> ScreenCallRegex.containsMatchIn(adapter.body) }
            }
        }
    }

    private companion object {
        private val ScreenCallRegex = Regex("""\b[A-Z]\w*Screen\(""")

        private val Message = LintRuleMessage(
            rule = "Every Adapter delegates rendering to a Screen composable.",
            why = "The Adapter/Screen split only works if the Adapter is a thin bridge: collect state, bind callbacks, render the Screen.",
            howToFix = "Move layout code into a *Screen composable and call it from the Adapter.",
            badExample = "@Composable fun GridAdapter(viewModel: GridViewModel = koinViewModel()) { TemplateGrid(...) { ... } }",
            goodExample = "@Composable fun GridAdapter(viewModel: GridViewModel = koinViewModel()) { GridScreen(state = state, onCellTap = ...) }",
        )
    }
}

File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/UI/AdaptersCallScreens.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class AdaptersCallScreens: QuickSpec {
    override class func spec() {
        Given("A feature Adapter view") {
            let adapters = WoofHarmonize.adapterViews.withoutName("ErrorAdapter")

            Then("It renders a Screen view") {
                adapters.assertTrue(message: message) { $0.description.containsMatch(of: screenCallPattern) }
            }
        }
    }

    private static let screenCallPattern = try! NSRegularExpression(pattern: #"\b[A-Z]\w*Screen\("#)

    private static let message = LintRuleMessage(
        rule: "Every Adapter delegates rendering to a Screen view.",
        why: "The Adapter/Screen split only works if the Adapter is a thin bridge: observe state, bind callbacks, render the Screen.",
        howToFix: "Move layout code into a *Screen view and call it from the Adapter.",
        badExample: "struct GridAdapter: View { var body: some View { TemplateGrid(...) { ... } } }",
        goodExample: "struct GridAdapter: View { var body: some View { GridScreen(state: viewModel.state, onCellTap: ...) } }"
    )
}
